### PR TITLE
Add Template Expressions

### PR DESCRIPTION
**Pull request by mattnibs**
Add the ability to include template expressions inside of
StringLiterals.

### CODE REVIEW

**Comment by mattnibs** [comment]
Closes #3343

File: compiler/parser/ztests/string-templates.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+   "hello 776f726c64"
+   "hello ${this}"
+   "hello ${to_hex(this)}"
+   "==="
+   "incompatible types"(error)
+   "hello 7f000001"
+   "hello ${this}"
+   "hello ${to_hex(this)}"
+   "==="
+
```
Comment (nwt) [inline compiler/parser/ztests/string-templates.yaml:24]:
```suggestion
```

File: compiler/parser/ztests/template-literals.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+   "hello 776f726c64"
+   "hello ${this}"
+   "hello ${to_hex(this)}"
+   "yo dawg world"
+   "==="
+   "incompatible types"(error)
+   "hello 7f000001"
+   "hello ${this}"
+   "hello ${to_hex(this)}"
+   "yo incompatible types"(bstring)
```
Comment (mattnibs) [inline compiler/parser/ztests/template-literals.yaml:25]: This feels like a bug, but not from this pr.

Comment (mccanne) [reply compiler/parser/ztests/template-literals.yaml:25]: > This feels like a bug, but not from this pr.

This should be fixed in #3396

File: compiler/parser/support.go [hunk 3]
```diff
@@ -33,6 +33,20 @@ func makeArgMap(args interface{}) (interface{}, error) {
 	return m, nil
 }
 
+func makeTemplateExprChain(in interface{}) interface{} {
```
Comment (nwt) [inline compiler/parser/support.go:36]: Maybe someday we can get rid of this and just use makeBinaryExprChain since that's so similar.

**Review by nwt — APPROVED** [review]

**Review by nwt — APPROVED** [review]

**Pull request merged**